### PR TITLE
Better communication computation overlap

### PR DESCRIPTION
**Pull request by snarayan21**
This PR patches torch to get better communication/computation overlap. Instead of using just one stream to unshard (all-gather) both attention params and FFN params, this uses a separate unshard stream for each. Computation now does not have to unnecessarily wait on communication from unrelated parts of the network.

Testing shows similar convergence with and without patching. Training instabilities, mainly due to load-balancing loss and not shuffling, were observed both with and without patching, which impacted convergence nondeterministically.
Screenshot 2023-12-26 at 10 30 25 PM
Screenshot 2023-12-28 at 11 06 02 PM

MFU improves by ~1-2% at small scales (first image) and ~25% at large scales, where without the patch, MFU is ~31% - (second image).
Screenshot 2024-01-02 at 4 36 46 PM
image

See this PR to PyTorch for more context: https://github.com/pytorch/pytorch/pull/116611

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/trainer/mosaic_fsdp.py [hunk 1]
```diff
@@ -54,9 +54,17 @@ def patch_pytorch():
         from torch.distributed.fsdp import _runtime_utils
         _runtime_utils._validate_and_get_hybrid_shard_state = lambda *args, **kwargs: None
 
+        # Better overlap communication and computation
+        from composer.trainer.mosaic_fsdp_utils import new_share_state_and_init_handle_attrs
+        _runtime_utils._share_state_and_init_handle_attrs = new_share_state_and_init_handle_attrs
+
```
Comment (mvpatel2000) [inline composer/trainer/mosaic_fsdp.py:60]: Let's remove for 2.1.0

**Review by mvpatel2000 — COMMENTED** [review]
Can you comment the monkeypatch block please? 

LGTM otherwise mod test failures (import gating?)

File: composer/trainer/mosaic_fsdp.py (hunk 1 above)
Comment (snarayan21) [reply composer/trainer/mosaic_fsdp.py]: done

**Comment by snarayan21** [comment]
I did have comments on the monkeypatched portion of the function, are there other comments you'd like me to add?

**Review by mvpatel2000 — APPROVED** [review]
🤦 missed comment, was on phone. LGTM!!
